### ISSUE TITLE
Invoking a service yields 2 issues -- (a) incorrect data is received by the service handler on the MCU, and (b) no response is received by the caller

### ISSUE THREAD

**Issue by safdark** [micro-ROS/micro_ros_arduino#750]
I've looked through the previous issues regarding service invocations, and incorporated the suggestions into my code, but I still face issues with service invocation. I am trying to use a precompiled micro_ros_arduino library in platformio (obtained from the below repository):
```
https://github.com/DominikN/micro_ros_arduino.git
(branch: galactic_custom_build)
```

I'm hoping for one of the following:
(a) help with what I might be doing wrong in the node.cpp code below, or
(b) which other micro_ros_arduino library/repository I should include in platformio.ini that would work on an ESP32 MCU. (The original one at github.com/micro-ROS/micro_ros_arduino.git does not work for the platformio + esp32 + arduino combination).

Thanks in advance for your help!

## Issue template

- Hardware description: Ubuntu 20.04 (Agent), ESP32 (Client)
- RTOS: Arduino
- Installation type: Platformio
- Version or commit hash: https://github.com/DominikN/micro_ros_arduino.git#galactic_custom_build

#### Steps to reproduce the issue
I've implemented a node with just one service ("/control") handler:
```
node
  /platformio.ini
  /src
    /node.cpp
```

Platformio.ini file:
```
[env]
platform = espressif32
framework = arduino

[env:esp32_node]
board = esp32dev
monitor_speed = 115200
upload_speed = 921600
src_filter = +<node.cpp>
lib_deps =
	https://github.com/DominikN/micro_ros_arduino#galactic-custom-build
build_flags = 
    -L $PROJECT_DIR/.pio/libdeps/$PIOENV/micro_ros_arduino/src/esp32_5_2_0/
    -l microros
```

Node.cpp file:
```cpp
#include <micro_ros_arduino.h>
#include <micro_ros_utilities/string_utilities.h>
#include <rcl/error_handling.h>
#include <rcl/rcl.h>
#include <rclc/executor.h>
#include <rclc/rclc.h>
#include <std_msgs/msg/string.h>
#include <stdio.h>

#include <test_msgs/srv/basic_types.h>

// ------------------
// Setup wifi
// ------------------
const char *ssid = "........";          // <<<<<< FILL >>>>>>>
const char *password = ".........";     // <<<<<< FILL >>>>>>>
const char* agent_ip = ".........";     // <<<<<< FILL >>>>>>>
const uint agent_port = ".........";    // <<<<<< FILL >>>>>>>

// ----------------------
// Error handling macros
// ----------------------
#define RCCHECK(fn, error_handler)    \
  {                                   \
    rcl_ret_t temp_rc = fn;           \
    if ((temp_rc != RCL_RET_OK)) {    \
      Serial.printf("(!=%d) [line %d] [%s]", temp_rc, __LINE__, rcutils_get_error_string().str); \
      error_handler;                  \
    }                                 \
  }

#define RCCRETRY(fn)                  \
  {                                   \
    rcl_ret_t temp_rc;                \
    do {                              \
      temp_rc = fn;                   \
    } while (temp_rc != RCL_RET_OK);  \
  }

#define RCSOFTCHECK(fn)               \
  {                                   \
    rcl_ret_t temp_rc = fn;           \
    if ((temp_rc != RCL_RET_OK)) {    \
      Serial.printf("(!=%d) [line %d] [%s]", temp_rc, __LINE__, rcutils_get_error_string().str); \
    }                                 \
  }


// --------------------
// PIN initialization
// --------------------
#define LED_PIN 2


// --------------------
// Node initialization
// --------------------
bool uros_initialized = false;
rcl_node_t node;
rclc_support_t support;
rcl_init_options_t init_options;
rcl_allocator_t allocator;
rclc_executor_t executor;
int num_execution_handles = 1; // 1 service handler

// ------------------
// Service objects
// ------------------
rcl_service_t service;
test_msgs__srv__BasicTypes_Request request;
test_msgs__srv__BasicTypes_Response response;

// ------------------
// Teardown
// ------------------
void teardown_ros() {
  if (uros_initialized) {
    rcl_ret_t rc = rcl_service_fini(&service, &node);
    rc += rcl_node_fini(&node);
    rc += rclc_executor_fini(&executor);
    rc += rclc_support_fini(&support);
    test_msgs__srv__BasicTypes_Request__fini(&request);
    test_msgs__srv__BasicTypes_Response__fini(&response);
    if (rc != RCL_RET_OK) {
      Serial.println("Error while cleaning up!");
    }
    uros_initialized = false;
  }
}

// ------------------
// Error loop
// ------------------
void error_loop() {
  teardown_ros();
  while (1) {
    Serial.println("error loop");
    delay(200);
    digitalWrite(LED_PIN, !digitalRead(LED_PIN));
  }
}

// ------------------
// Control service handler
// ------------------
void service_callback(const void * req_in, void * res_in) {
  Serial.print("<Service Invocation callback>");
  test_msgs__srv__BasicTypes_Request * _request = (test_msgs__srv__BasicTypes_Request *) req_in;
  test_msgs__srv__BasicTypes_Response * _response = (test_msgs__srv__BasicTypes_Response *) res_in;
  _response->int16_value = _request->int16_value - 1;
  Serial.printf("Control/State: %d / %d~\n", (int) _request->int16_value, (int) _response->int16_value);
}

// ------------------
// Initialization
// ------------------
void initialize_ros() {
  Serial.print("Initializing ros");
  set_microros_wifi_transports((char*)ssid, (char*)password, (char*)agent_ip, (uint)agent_port);
  
  // create allocator
  Serial.print(".");
  allocator = rcl_get_default_allocator();

  // create init_options
  Serial.print(".");
  RCCHECK(rclc_support_init(&support, 0, NULL, &allocator), error_loop());

  // setup node
  Serial.print(".");
  node = rcl_get_zero_initialized_node();
  Serial.print(".");
  RCCHECK(rclc_node_init_default(&node, "control_node", "arm/forearm", &support), error_loop());

  // setup executor
  Serial.print(".");
  executor = rclc_executor_get_zero_initialized_executor();
  Serial.print(".");
  RCCHECK(rclc_executor_init(&executor, &support.context, num_execution_handles, &allocator), error_loop());
  Serial.print(".");
  RCCHECK(rclc_executor_set_timeout(&executor, RCL_MS_TO_NS(2000)), error_loop());

  // setup service
  Serial.print(".");
  test_msgs__srv__BasicTypes_Request__init(&request);
  Serial.print(".");
  test_msgs__srv__BasicTypes_Response__init(&response);
  Serial.print(".");
  RCCHECK(rclc_service_init_default(&service, &node, ROSIDL_GET_SRV_TYPE_SUPPORT(test_msgs, srv, BasicTypes), "/control"), error_loop());
  Serial.print(".");
  RCCHECK(rclc_executor_add_service(&executor, &service, (void*) &request, (void*) &response, &service_callback), error_loop());

  Serial.print(".");

  uros_initialized = true;

  Serial.println("Done");
}

// ------------------
// Responsiveness variables
// ------------------
int loop_frequency = 2;                        // Hz


// ------------------
// Setup()
// ------------------
void setup() {
  // Setup pins
  pinMode(LED_PIN, OUTPUT);
  Serial.begin(115200);

  Serial.println("Initializing:");
  delay(3000);

  initialize_ros();

  digitalWrite(LED_PIN, HIGH);
  Serial.println("Startup complete!");
  Serial.print("[Looping]");
}

// ------------------
// Loop()
// ------------------
void loop() {
  int delay_ms = (1.0 / loop_frequency) * 1000;
  delay(delay_ms);

  Serial.print("\n. ");
  RCSOFTCHECK(rclc_executor_spin_some(&executor, RCL_MS_TO_NS(100)));
  digitalWrite(LED_PIN, !digitalRead(LED_PIN));
}
```

Run the agent with:
```
ros2 run micro_ros_agent micro_ros_agent udp4 -p 8888 -v 5
```

Flash the MCU via platformio:
```
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0018,len:4
load:0x3fff001c,len:1044
load:0x40078000,len:10124
load:0x40080400,len:5828
entry 0x400806a8
Initializing:
Initializing ros............Done
Startup complete!
[Looping]
. 
. 
<...>
```

Issue the service call:
```
ros2 service call /control test_msgs/srv/BasicTypes "{int16_value: 5}"
```

#### Expected behavior

The MCU's service handler should receive the right values sent by the caller, and should return with the response to the caller.

#### Actual behavior

The request message received by the service handler doesn't have the right value set ("int16_value"), and the response is never returned to the caller.

The client seems to have registered appropriately with the agent:
```
[1644217216.858359] info     | UDPv4AgentLinux.cpp | init                     | running...             | port: 8888
[1644217216.858487] info     | Root.cpp           | set_verbose_level        | logger setup           | verbose_level: 5
[1644217224.160057] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0x00000000, len: 24
[1644217224.160361] info     | Root.cpp           | create_client            | create                 | client_key: 0x3AB0877C, session_id: 0x81
[1644217224.160507] info     | SessionManager.hpp | establish_session        | session established    | client_key: 0x3AB0877C, address: 192.168.1.44:47138
[1644217224.160709] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0x3AB0877C, len: 19
[1644217224.171900] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0x3AB0877C, len: 56
[1644217224.200179] info     | ProxyClient.cpp    | create_participant       | participant created    | client_key: 0x3AB0877C, participant_id: 0x000(1)
[1644217224.200383] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0x3AB0877C, len: 14
[1644217224.200423] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0x3AB0877C, len: 13
[1644217224.230011] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0x3AB0877C, len: 13
[1644217224.230437] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0x3AB0877C, len: 180
[1644217224.231780] info     | ProxyClient.cpp    | create_replier           | replier created        | client_key: 0x3AB0877C, requester_id: 0x000(7), participant_id: 0x000(1)
[1644217224.231888] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0x3AB0877C, len: 14
[1644217224.231945] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0x3AB0877C, len: 13
[1644217224.238551] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0x3AB0877C, len: 13
[1644217224.760102] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0x3AB0877C, len: 24
[1644217224.760188] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0x3AB0877C, len: 13
[1644217224.760383] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0x3AB0877C, len: 13
[1644217224.760532] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0x3AB0877C, len: 13
[1644217224.771841] debug    | UDPv4AgentLinux.cpp | recv_message             | [==>> UDP <<==]        | client_key: 0x3AB0877C, len: 24
[1644217224.772136] debug    | UDPv4AgentLinux.cpp | send_message             | [** <<UDP>> **]        | client_key: 0x3AB0877C, len: 13
```

But, when I call the '/control' service via the command line, like so:
```
ros2 service call /control test_msgs/srv/BasicTypes "{int16_value: 5}"
```
... I get:
```
1644215962.774168 [0]       ros2: using network interface eno1 (udp/192.168.1.100) selected arbitrarily from: eno1, docker0
requester: making request: test_msgs.srv.BasicTypes_Request(bool_value=False, byte_value=b'\x00', char_value=0, float32_value=0.0, float64_value=0.0, int8_value=0, uint8_value=0, int16_value=5, uint16_value=0, int32_value=0, uint32_value=0, int64_value=0, uint64_value=0, string_value='')
```
... which never returns.

The service handler on the client gets invoked with the wrong request values (int16_value = 0 instead of 5):
```
. <Service Invocation callback>Control/State: 0 / -1~
.
.
.
<...>
```

The loop() continues on the MCU, but the caller never gets a response from the service.
```
.
.
.
.
<...>
```

#### Additional information

I can repeatedly issue the service calls. They are received on the MCU, but the issues above are consistent (missing data in the request message, and no response received by the caller).

```
. <Service Invocation callback>Control/State: 0 / -1~
.
.
. <Service Invocation callback>Control/State: 0 / -1~
.
.
.
. <Service Invocation callback>Control/State: 0 / -1~
```

**Comment by pablogs9** [comment]
On the ROS 2 side are you using Fast DDS as middleware? In ROS 2 galactic there is another default middleware and micro-ROS is only guaranteed to work with Fast DDS: https://docs.ros.org/en/galactic/Installation/DDS-Implementations/Working-with-eProsima-Fast-DDS.html

**Comment by safdark** [comment]
Thanks @pablogs9! That worked perfectly! Unless I've overlooked that detail somewhere, it would be great if this could be documented in the installation section for micro-ROS, since (as you said) the default RMW in galactic is no longer the one that micro-ROS is compatible with.
Thanks again for the prompt reply!

**Comment by safdark** [comment]
@pablogs9 ... I did some more testing (after enabling Fast DDS), and am seeing now that though the service calls do generally return, sometimes the invocations are not even received on the MCU (and the call just hangs). So, the issue from before was resolved (in that the MCU now responds to any service call it receives), but this new issue has surfaced. Any thoughts?

**Comment by safdark** [comment]
There's no option for me to reopen the issue from this end.

**Comment by pablogs9** [comment]
You are not supposed to use `std_msgs__msg__String__init`-like functions in micro-ROS. For example in this case, reallocations are expected to happen when writing the string.

You have to provide memory to your data types in micro-ROS. Please read [this tutorial](https://micro.ros.org/docs/tutorials/advanced/handling_type_memory/). And check [this package](https://github.com/micro-ROS/micro_ros_utilities).

Let us know if this solves the issue.

**Comment by safdark** [comment]
Thanks @pablogs9 ! I got this working for initializing the std_msgs/msg/String type (but now the service handler doesn't respond again). So, I presume similar memory allocation is also needed for the service messages? In my case, I use the test_msgs.srv.BasicTypes service interface, for which I define two message objects:
```
test_msgs__srv__BasicTypes_Request request;
test_msgs__srv__BasicTypes_Response response;
```

But I don't see any APIs that seem intuitively matched to the *service* message types, for memory allocation (other than *test_msgs__srv__BasicTypes_Request__init* and *test_msgs__srv__BasicTypes_Response__init*, which you said should not be used in Galactic). Perhaps you could provide some sample code for allocating memory for the above *service* messages?

**Comment by pablogs9** [comment]
Is it possible for you to use this repo instead of `galactic_custom_build` of `https://github.com/DominikN/micro_ros_arduino.git`?

**Comment by safdark** [comment]
I did try that earlier, but get a linker error:
```
> Executing task: platformio run --environment esp32_node <

Processing esp32_node (board: esp32dev; platform: espressif32; framework: arduino)
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (3.5.0) > Espressif ESP32 Dev Module
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (esp-prog) External (esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-arduinoespressif32 3.10006.210326 (1.0.6) 
 - tool-esptoolpy 1.30100.210531 (3.1.0) 
 - toolchain-xtensa32 2.50200.97 (5.2.0)
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 29 compatible libraries
Scanning dependencies...
Dependency Graph
|-- <micro_ros_arduino> 0.0.4+sha.d63e576
|   |-- <SPI> 1.0
|   |-- <WiFi> 1.0
|   |-- <ESP32 BLE Arduino> 1.0.1
Building in release mode
Linking .pio/build/esp32_node/firmware.elf
/home/safdar/.platformio/packages/toolchain-xtensa32/bin/../lib/gcc/xtensa-esp32-elf/5.2.0/../../../../xtensa-esp32-elf/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
*** [.pio/build/esp32_node/firmware.elf] Error 1
```

Here's my updated platformio.ini:
```
[env]
platform = espressif32
framework = arduino

[env:esp32_node]
board = esp32dev
monitor_speed = 115200
upload_speed = 921600
src_filter = +<node.cpp>
lib_deps =
    https://github.com/micro-ROS/micro_ros_arduino
build_flags = 
    -L $PROJECT_DIR/.pio/libdeps/$PIOENV/micro_ros_arduino/src/esp32/
    -l microros
```

**Comment by pablogs9** [comment]
@Acuadros95 any idea on this linking issue? Do you know how to handle the ESP32 compilers in order to use the library with micro-ROS repo?

@safdark are you using Arduino/Platform.IO for any reason? When using ESP32 we recommend using its development environment (ESP-IDF) and integrating micro-ROS with [micro_ros_espidf_component](https://github.com/micro-ROS/micro_ros_espidf_component) that is well-tested and works smotthly.

**Comment by safdark** [comment]
I will check out the micro_ros_espidf_component. I'm new to embedded systems development, so unless tutorials are detailed, I can feel a bit lost ;). In the morning I will give that a try. If you know a detailed article that walks one through setting up the esp-idf component and integration with micro-ROS (assuming no prior familiarity with embedded systems), I'd much appreciate a link to it.

Meanwhile, I do have some Teensy boards as well, and had punted on that for the time being since Teensy's support for certain motor drivers seemed limited (the last time I checked). My end goal is to run various consumer-grade stepper and servo motors from the MCU, using the most commonly available/standard motor driver libraries. I'd be open to trying out a Teensy/Arduino combo, if that is *better* integrated with micro-ros/arduino/platformio. I'm not familiar with FreeRTOS yet ... hence the decision to stick with Arduino. Platformio's accessibility and simplicity for a layman, and its flexibility to switch between different boards, libraries and platforms was why I chose it.

**Comment by Acuadros95** [comment]
That linking bug is not related to micro-ROS, but as a work arround:
```
[env:esp32dev]
platform = https://github.com/platformio/platform-espressif32.git#feature/arduino-upstream
board = esp32dev
framework = arduino
lib_deps = 
    https://github.com/micro-ROS/micro_ros_arduino.git
build_flags =
    -L ./.pio/libdeps/esp32dev/micro_ros_arduino/src/esp32/
    -l microros
    -D ESP32

platform_packages =
  toolchain-xtensa32 @ ~2.80400.0
  framework-arduinoespressif32@https://github.com/espressif/arduino-esp32.git#2.0.2
```

This issue was resolved here: https://github.com/micro-ROS/micro_ros_arduino/issues/736

**Comment by pablogs9** [comment]
@safdark You have a bunch of documentation on ESP-IDF in their [webpage](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/). Including micro-ROS in an ESP-IDF project is simple, you have instructions in our [repo](https://github.com/micro-ROS/micro_ros_espidf_component#micro-ros-component-for-esp-idf). Open issues there if you have any questions.

**Comment by safdark** [comment]
Thanks @Acuadros95 and @pablogs9 .

@Acuadros95 , I tried out the platformio.ini file you listed, and the build compiles and links now. **However**, the service handler on the server is still *intermittently* unresponsive. It is a slight improvement from the previous case, in that at least the service responds a couple of times, but overall, it is still unusable for the above reason. For example, if I issue 6 service calls, ~50% of the calls consistently hang without a response (and the service handler on the MCU doesn't get invoked in those cases). So, as it stands, I am blocked on moving forward with this particular framework/platform/board combination :(.

@pablogs9 Yes, I will switch to using freertos and esp-idf, and see if I have better luck there.

Thanks again for both your help.

**Comment by pablogs9** [comment]
Can you share the agent output (with the flag `-v6`) of a normal operation of your setup. Preferably indicating when you are launching a service that is ignored.

**Comment by pablogs9** [comment]
Here you have a basic test of your app in our ESP32. No hangs. Have you checked your network?

https://drive.google.com/file/d/1y7e5yruNDLzbIcAv6m7hBrt9TiCoreAD/view?usp=sharing

**Comment by safdark** [comment]
That is interesting, and reassuring!!! Thanks so much for the video!

I will check my wifi connection soon, and report back with what I find.

Much appreciated, @pablogs9 :)

**Comment by safdark** [comment]
@pablogs9 and @Acuadros95 ... the issue was indeed with the network. I switched to another modality and am not facing this issue now. So, the access point I was previously using between the agent and client was likely the culprit. It is possible the intermittency with sending the service calls (despite all the code fixes) could have been due to using *udp* to communicate over that crappy access point. That might explain why the service invocations never reached the client, though I'd have hoped that a failed invocation would have shown an error on the caller side (ros2 service run ...). Given the inner workings of DDS, can a service caller programmatically detect if an invocation did indeed fail over an unreliable medium between the agent and client? If not, could this be something that a diagnostic tool like *ros wtf* (or other) could have detected earlier on as part of the investigation?

My apologies for the false alarm, and the trouble! And many thanks again for all the help!
